Issues with interactions with custom Scanners

I am working on something involving this database driver that makes use of the sql.Null\* types. When making queries that have a mix of NULL and non-NULL values in a column, only NULL values are returned. Consider this SQLite shell session:

```
sqlite> CREATE TABLE foo ( data TEXT );
sqlite> INSERT INTO foo(data) VALUES ('bar');
sqlite> INSERT INTO foo(data) VALUES (NULL);
```

When I run this in Go and query `SELECT data FROM foo ORDER BY rowid`, I get the following (after scanning it into a slice of sql.NullString):

```go
[]sql.NullString{
  {String: "bar", Valid: true},
  {String: "", Valid: true},
}
```

Which is confusing because the SQLite shell returns the following for the moral equivalent of that datatype:

```
sqlite> SELECT data, data IS NOT NULL AS valid FROM foo ORDER BY rowid;
| data | valid |
|------|-------|
| bar  | 1     |
|      | 0     |
```

Here is a minimal reproduction case:

```go
func TestSQLiteBug(t *testing.T) {
	db := sql.OpenDB(sqlite.Connector(":memory:", func(ctx context.Context, conn driver.ConnPrepareContext) error { return nil }, nil))
	err := db.Ping()
	if err != nil {
		t.Fatal(err)
	}

	_, err = db.Exec("CREATE TABLE foo ( data TEXT )")
	if err != nil {
		t.Fatal(err)
	}

	db.Exec("INSERT INTO foo(data) VALUES ('bar')")
	db.Exec("INSERT INTO foo(data) VALUES (NULL)")

	var result []sql.NullString
	rows, err := db.Query("SELECT data FROM foo ORDER BY rowid")
	if err != nil {
		t.Fatal(err)
	}
	defer rows.Close()

	for rows.Next() {
		var val sql.NullString
		err := rows.Scan(&val)
		if err != nil {
			t.Fatal(err)
		}
		result = append(result, val)
	}

	if len(result) != 2 {
		t.Fatalf("wanted 2 results, not %d", len(result))
	}

	first := result[0]
	second := result[1]

	t.Logf("first:  %#+v", first)
	t.Logf("second: %#+v", second)

	if !first.Valid {
		t.Error("first is not valid but we wanted it to be")
	}

	if first.String != "bar" {
		t.Errorf("first is not \"bar\", got: %q", first.String)
	}

	if second.Valid {
		t.Error("second is valid but we wanted it to not be")
	}
}
```

This fails with the following message:

```
=== RUN   TestSQLiteBug
    sqlite_date_test.go:101: first:  sql.NullString{String:"bar", Valid:true}
    sqlite_date_test.go:102: second: sql.NullString{String:"", Valid:true}
    sqlite_date_test.go:113: second is valid but we wanted it to not be
--- FAIL: TestSQLiteBug (0.00s)
```

There have been other issues when adding left joins into the mix, but details will be provided in the comments when I am able to get a minimal reproduction.

--- a/sqlite.go
+++ b/sqlite.go
@@ -374,7 +374,6 @@ type stmt struct {
 	prepCtx context.Context // the context provided to prepare, for tracing
 
 	// filled on first step only if persist==true
-	colTypes     []sqliteh.ColumnType
 	colDeclTypes []colDeclType
 	colNames     []string
 }
@@ -634,7 +633,6 @@ type rows struct {
 	colNames []string // filled on call to Columns
 
 	// Filled on first call to Next.
-	colTypes     []sqliteh.ColumnType
 	colDeclTypes []colDeclType
 }
 
@@ -681,28 +679,24 @@ func (r *rows) Next(dest []driver.Value) error {
 		return io.EOF
 	}
 
-	if r.colTypes == nil {
-		if r.stmt.colTypes != nil {
-			r.colTypes = r.stmt.colTypes
-			r.colDeclTypes = r.stmt.colDeclTypes
-		} else {
-			colCount := r.stmt.stmt.ColumnCount()
-			r.colTypes = make([]sqliteh.ColumnType, colCount)
-			r.colDeclTypes = make([]colDeclType, colCount)
-			for i := range r.colTypes {
-				r.colTypes[i] = r.stmt.stmt.ColumnType(i)
-				r.colDeclTypes[i] = colDeclTypeFromString(r.stmt.stmt.ColumnDeclType(i))
-			}
-			if r.stmt.persist {
-				r.stmt.colTypes = r.colTypes
-				r.stmt.colDeclTypes = r.colDeclTypes
-			}
+	if r.colDeclTypes == nil {
+		r.colDeclTypes = r.stmt.colDeclTypes
+	}
+	if r.colDeclTypes == nil {
+		colCount := r.stmt.stmt.ColumnCount()
+		r.colDeclTypes = make([]colDeclType, colCount)
+		for i := range r.colDeclTypes {
+			r.colDeclTypes[i] = colDeclTypeFromString(r.stmt.stmt.ColumnDeclType(i))
+		}
+		if r.stmt.persist {
+			r.stmt.colDeclTypes = r.colDeclTypes
 		}
 	}
 
 	for i := range dest {
+		colType := r.stmt.stmt.ColumnType(i)
 		if r.colDeclTypes[i] == declTypeDateOrTime {
-			switch r.colTypes[i] {
+			switch colType {
 			case sqliteh.SQLITE_INTEGER:
 				v := r.stmt.stmt.ColumnInt64(i)
 				dest[i] = time.Unix(v, 0)
@@ -729,7 +723,7 @@ func (r *rows) Next(dest []driver.Value) error {
 			}
 			continue
 		}
-		switch r.colTypes[i] {
+		switch colType {
 		case sqliteh.SQLITE_INTEGER:
 			val := r.stmt.stmt.ColumnInt64(i)
 			if r.colDeclTypes[i] == declTypeBoolean {

--- a/sqlite_test.go
+++ b/sqlite_test.go
@@ -857,6 +857,88 @@ func TestSetWALHook(t *testing.T) {
 	}
 }
 
+// Tests that we don't remember the SQLite column types of the first row of the
+// result set (notably the "NULL" type) and re-use it for all subsequent rows
+// like we used to.
+func TestNoStickyColumnTypes(t *testing.T) {
+	db := openTestDB(t)
+	exec(t, db, "CREATE TABLE t (id INTEGER PRIMARY KEY, v1 ANY, v2 ANY)")
+
+	type row []any
+	r := func(v ...any) row { return v }
+	rs := func(v ...row) []row { return v }
+	tests := []struct {
+		name string
+		rows []row
+	}{
+		{"no-null", rs(
+			r("a", "b"),
+			r("foo", "bar"))},
+		{"only-null", rs(
+			r(nil, nil))},
+		{"null-after-string", rs(
+			r("a", "b"),
+			r(nil, "bar"))},
+		{"string-after-null", rs(
+			r(nil, "b"),
+			r("foo", "bar"))},
+		{"null-after-int", rs(
+			r(101, 102),
+			r(nil, 202))},
+		{"int-after-null", rs(
+			r(nil, 102),
+			r(201, 202))},
+		{"changing-types-within-a-column-between-rows", rs(
+			r("foo", nil),
+			r(nil, 2),
+			r(3, "bar"))},
+	}
+
+	// canonical maps from types we get back out from sqlite
+	// to the types we provided in the test cases above.
+	canonical := func(v any) any {
+		switch v := v.(type) {
+		default:
+			return v
+		case []byte:
+			return string(v)
+		case int64:
+			return int(v)
+		}
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			exec(t, db, "DELETE FROM t")
+			for primaryKey, r := range tt.rows {
+				exec(t, db, "INSERT INTO t VALUES (?, ?, ?)", append([]any{primaryKey}, r...)...)
+			}
+			rows, err := db.Query("SELECT id, v1, v2 FROM t ORDER BY id")
+			if err != nil {
+				t.Fatal(err)
+			}
+			for rows.Next() {
+				var id int
+				var v1, v2 any
+				err := rows.Scan(&id, &v1, &v2)
+				if err != nil {
+					t.Fatal(err)
+				}
+				v1, v2 = canonical(v1), canonical(v2)
+				want := tt.rows[id]
+				got := row{v1, v2}
+				t.Logf("[%v]: %T, %T", id, v1, v2)
+				if !reflect.DeepEqual(got, want) {
+					t.Errorf("row %d got %v; want %v", id, got, want)
+				}
+			}
+			if err := rows.Err(); err != nil {
+				t.Fatal(err)
+			}
+		})
+	}
+}
+
 func BenchmarkWALHookAndExec(b *testing.B) {
 	ctx := context.Background()
 	db := openTestDB(b)

